File: 2024 May/Day 3: 165. Compare Version Numbers [Medium]/sol.py
class Solution:
    def compareVersion(self, version1: str, version2: str) -> int:
        v1=version1.split(".")
        v2=version2.split(".")
        l1=len(v1)-1
        l2=len(v2)-1

        while v1!=[] and int(v1[l1])==0:
            v1.pop()
            l1-=1
        while v2!=[] and int(v2[l2])==0:
            v2.pop()
            l2-=1
        
        for i in range(min(len(v1),len(v2))):
            if int(v1[i])>int(v2[i]):
                return 1
            elif int(v1[i])<int(v2[i]):
                return -1
        
        if len(v1)>len(v2):
            return 1
        elif len(v1)<len(v2):
            return -1
        
        return 0
    
# Python    
# 28 ms (beats 91.52%)
# 16.59 mb (beats 47.04%)

File: solutions/857 Minimum Cost to Hire K Workers [Hard]/sol.cpp
#pragma GCC optimize("O3", "unroll-loops")

class Solution {
public:
    double mincostToHireWorkers(vector<int>& quality, vector<int>& wage, int k) {
        int n = quality.size();
        vector<array<double,2>> worker(n); // wage/quality, quality
        for(int i=0; i<n; i++) worker[i] = {(double)wage[i]/quality[i],(double)quality[i]}; 

        sort(worker.begin(),worker.end());

        int qualitysum = 0;

        priority_queue<int> pq; // maxheap (quality)
        for(int i=0; i<k; i++){ // k best (quality/wage)
            qualitysum += worker[i][1];
            pq.push(worker[i][1]);
        }
        double pay = qualitysum * worker[k-1][0];
        for(int i=k; i<n; i++) {
            pq.push(worker[i][1]);
            qualitysum += worker[i][1]-pq.top();
            pq.pop();
            pay = min(pay,qualitysum*worker[i][0]);
        }
        
        return pay;
    }
};

const int ZERO = []()
{
	std::ios_base::sync_with_stdio(false);
	std::cin.tie(nullptr);
	return 0;
}();


// C++
// 15 ms (beats 94.64%)
// 24.97 mb (beats 84.95%)